scheduler/src/cook/mesos/monitor.clj

```diff
@@ -89,27 +83,61 @@
            (if (every? true? (map (fn [[resource amount]]
                                     (< (or (resource used-resources) 0.0)
                                        amount))
-                                  promised-resources))
+                                  (promised-resources user)))
              (recur users (assoc starved-stats user (compute-starvation user)))
              (recur users starved-stats)))
          starved-stats)))))
 
-(defn make-stats-events
-  "Return a list of riemann events for jobs with the given state, e.g. running,
-   waiting and starved. These events can be used by (riemann/send-events ...)."
-  [db state stats]
-  (->> stats
-       (add-aggregated-stats db)
-       (mapcat (fn [[user stats]]
-                 (map (fn [[type amount]]
-                        (assoc event
-                               :service (apply format "cook scheduler %s %s %s"
-                                               (map name [state user type]))
-                               :metric amount))
-                      stats)))))
+(defn- set-counter!
+  "Sets the value of the counter to the new value.
+   A data race is possible if two threads invoke this function concurrently."
+  [counter value]
+  (let [amount-to-inc (- (int value) (counters/value counter))]
+    (counters/inc! counter amount-to-inc)))
+
+(defn- clear-old-counters!
+  "Clears counters that were present on the previous iteration
+  but not in the current iteration. This avoids the situation
+  where a user's job changes state but the old state's counter
+  doesn't reflect the change."
+  [state stats state->previous-stats-atom]
+  (let [previous-stats (get @state->previous-stats-atom state)
+        previous-users (set (keys previous-stats))
+        current-users (set (keys stats))
+        users-to-clear (difference previous-users current-users)]
+    (run! (fn [user]
+            (run! (fn [[type _]]
+                    (-> [state user (name type)]
+                        counters/counter
+                        counters/clear!))
```

Comment:
Can we delete the counter instead of just clearing it?

Reply:
Done